placement/interface.go

```diff
@@ -0,0 +1,104 @@
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+// Algorithm places shards on hosts
+type Algorithm interface {
+	// InitPlacement initialize a sharding placement with RF = 1
+	BuildInitialPlacement(hosts []Host, ids []uint32) (Snapshot, error)
+
+	// AddReplica up the RF by 1 in the cluster
+	AddReplica(p Snapshot) (Snapshot, error)
+
+	// AddHost adds a host to the cluster
+	AddHost(p Snapshot, h Host) (Snapshot, error)
+
+	// RemoveHost removes a host from the cluster
+	RemoveHost(p Snapshot, h Host) (Snapshot, error)
+
+	// ReplaceHost replace a host with a new host
+	ReplaceHost(p Snapshot, leavingHost, addingHost Host) (Snapshot, error)
+}
+
+// DeploymentPlanner generates deployment steps for a placement
+type DeploymentPlanner interface {
+	// DeploymentSteps returns the deployment steps
+	DeploymentSteps(p Snapshot) [][]HostShards
+}
+
+// Snapshot describes how shards are placed on hosts
+type Snapshot interface {
+	// HostShards returns all HostShards in the placement
+	HostShards() []HostShards
+
+	// HostsLen returns the length of all hosts in the placement
+	HostsLen() int
+
+	// Replicas returns the replication factor in the placement
+	Replicas() int
+
+	// ShardsLen returns the number of shards in a replica
+	ShardsLen() int
+
+	// Shards returns all the unique shard ids for a replica
+	Shards() []uint32
+}
+
+// HostShards represents a host and its assigned shards
+type HostShards interface {
+	Host() Host
+	Shards() []uint32
+}
+
+// Host contains the information needed for placement
+type Host struct {
+	Rack    string
+	Address string
+}
+
+// Service handles the placement related operations for registered services
+// all write or update operations will persist the generated snapshot before returning success
+type Service interface {
+	BuildInitialPlacement(service string, hosts []string, shardLen int) error
+	AddReplica(service string) error
+	AddHost(service string, host string) error
+	AddHostFromPool(service string, pool string) error
+	RemoveHost(service string, host string) error
+	ReplaceHost(service string, leavingHostName string, addingHostName string) error
+	ReplaceHostFromPool(service string, leavingHostName string, pool string) error
+
+	// Snapshot gets the persisted snapshot for service
+	Snapshot(service string) (Snapshot, error)
+}
+
+// HostInventory provides inventory information
+type HostInventory interface {
```

Comment:
if we are removing the pool related methods from the Service interface, maybe we should remove the Dump(pool string) method here too? Also I don't want to merge this with the Host, Host is just a struct of 2 pieces of data (rack and address) it does not need to know where to retrieve the rack information for a host address?